Dynamic loading of OpenXR Loader on Android

kisg opened the pull request:
This change implements dynamic loading of the OpenXR Loader library on Android. All OpenXR symbols are also resolved at runtime using xrGetInstanceProcAddr.

On Windows and Linux the OpenXR loader is still included in the main engine binary.
On Android the loader is built as a separate shared library. This library is included in the custom build template, but currently not added to any of the build outputs.
An Android plugin may be used to add the required (and possibly device specific) OpenXR Loader library to the final APK.

Co-authored-by: @korompg

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

kisg commented:
@BastiaanOlij @m4gr3d @akien-mga This is the new PR I mentioned before.
A companion demo project with a Meta OpenXR Loader plugin may be found here: https://github.com/migeran/godot4_openxr_demo/tree/dynamic_openxr_loader

BastiaanOlij commented on modules/openxr/SCsub:
```diff
@@ -18,10 +18,12 @@ env_openxr.Prepend(
     ]
 )
 
+if env["platform"] == "android":
```
I would do the same here as with the other PR, just move the whole block forward so our OpenXR includes within the module source also resolve properly.

kisg replied:
Done.

BastiaanOlij commented:
So this replaces your old PR #64826 ?

This all looks really good, my only remark would be is that I would put all the function pointers and loader code into a separate class. Maybe we should create our own `openxr_loader.h/cpp` that has this logic and we could think about at some point also replacing the desktop code. But I may be overthinking things here :)

BastiaanOlij commented:
Ah off course that makes perfect sense. You're loading the loader which then loads the actual runtime, you're not loading the runtime itself so indeed very little changes. The only thing we're not using from the loader is the trampoline code and now implementing our own.

It's always felt a little ugly having all our own calling code intermingled with the actual implementation code. With extensions it was sort of ok because we're usually talking about 2 or 3 function calls to be added to the extension, but with the main code it feels like a lot of noise in OpenXRAPI. That said, it's not really a big deal and it also seems weird to have the functions related to extensions in the extensions, and the core functions in a separate file, there is something to say for consistency.

Generally speaking, the more we can implement in the extension classes the better, there are a few things were its just not practical to move everything into the extension classes because things are too tightly coupled to what is happening in the main classes. Especially around the action map and where certain structures need to be populated like the depth buffer stuff, there just wasn't a nice way to move that code into the extensions without jumping through major hoops, ideally though I'd like to find ways to clean this up and move more into the extension classes.

The other thing that we need to be mindful of, something I've discussed with Fredia in the past as well, is that we should minimize exposing platform dependent implementations to the end user. A good example of this is the depth buffer implementation where the official Khronos extension has one way of doing this, and Metas Spacewarp has a different way of doing this. That is one that is easily hidden from the end user but others aren't (platform differences in the hand tracking API is a good example here that we've not been able to overcome yet). On that subject, there is a PR outstanding for allowing us to tell Godot that we want depth and/or motion vector data that also implements Khronos' depth buffer API, but the mobile renderer is still missing a motion vector implementation. I've been looking into that but there are a few challenges yet to overcome.

Thanks for the headsup that you guys are looking into porting the Passthrough extension, I was planning on starting on that myself now that I have a bunch of other things finished. We almost doubled up there :) The pass through itself should be relatively straight forward to port, but in Godot 4 we do not have a settings object, properties are simply exposed on the XRInterface itself.

Are you guys on the [Godot Contributors Chat server](https://chat.godotengine.org/)? We have a dedicated XR channel there and it may be easier to converse in real time (impractical timezones not withstanding).

m4gr3d commented on platform/android/SCsub:
```diff
@@ -74,3 +74,10 @@ if lib_arch_dir != "":
     )
     env_android.Command(out_dir + "/libc++_shared.so", stl_lib_path, Copy("$TARGET", "$SOURCE"))
+
+    out_openxr_dir = "#platform/android/java/lib/libs/openxr/" + lib_type_dir + "/" + lib_arch_dir
```
Should this also be deleted in the `cleanGodotTemplates` task?

kisg replied:
This path is under lib/libs, which is already removed by cleanGodotTemplates.

kisg commented on platform/android/SCsub:
```diff
@@ -74,3 +74,10 @@ if lib_arch_dir != "":
+
+    out_openxr_dir = "#platform/android/java/lib/libs/openxr/" + lib_type_dir + "/" + lib_arch_dir
+    env_android.Command(
+        out_openxr_dir + "/libopenxr_loader.so",
```
Currently we only place these into the template, but they will not be added to the APK. When an AAR with an libopenxr_loader.so is added to the project, that will be included in the final APK. 

Unfortunately, due to the Android Gradle Plugin's limitation there is no safe way telling which .so should be added to the APK if multiple files with the same name are present during build. Here is an open issue about it in Google's tracker: https://issuetracker.google.com/issues/214395989

There are 2 options to solve this:
- Package the KHR loader that we build as a standard plugin that is included with the templates, so it can be easily added to the project. This is the option I would recommend.
- Hack the Gradle scripts e.g. to detect during the final copy that the OpenXR loader is missing from the APK and patch the APK.

I would support the first option is that a lot of Android projects are not AR/VR titles, so they don't need OpenXR anyways, and this way the APK will not include the additional ~500KB loader code.

What do you think?

BastiaanOlij replied:
I would not be against having the default Khronos loader downloadable as an AAR.

What I think is important is to make sure that we can deploy the Khronos loader, Meta loader and Pico loader together. I think this will work seeing we can turn on/off AARs in the export. This will enable people to create exports from the same project for different platforms.

kisg replied:
Should we generate an AAR + gdap file during the generateGodotTemplates task? If yes, can we do that in a follow-up PR, or do we need to do it in this one?

BastiaanOlij replied:
I'm on the fence about that one, @akien-mga what is your opinion on this?

@kisg btw did you get my DM on the Godot chat server? I shared a repo with you based on your Meta AAR with CI building both the AAR and a proper asset that can be uploaded to the asset library.

We could use that same approach for the Khronos loader.

m4gr3d replied:
I'm also in favor of the first option, so let's proceed with that.

> Should we generate an AAR + gdap file during the generateGodotTemplates task?

What's the behavior when no loader is provided? Is the logic smart enough to detect that and abort?

If it's, then the Khronos loader plugin should be generated separately from the Godot templates and provided as an asset as @BastiaanOlij mentioned alongside the other (e.g: Meta) loader plugins.

If the logic can't handle a missing loader, then it should be generated alongside the Godot template and provided as default.

akien-mga replied:
The first option sounds good to me too. It makes sense for both loaders to be provided as plugins, with Khronos being the one available out of the box since it's FOSS.

kisg replied:
@BastiaanOlij @m4gr3d @akien-mga It looks like we are in agreement here, thank you.

If you agree, we will add a new task ./gradlew generateKhronosLoaderOpenXRPlugin (or similar - please suggest a better name) that will generate the AAR and the gdap file in the bin folder.

BastiaanOlij replied:
I would just call it `generateHKROpenXRPlugin`. I'd also like the standardise the naming of the AARs so that if you have a couple you support they're nicely together in the plugins folder. 
So I called the meta one godot_openxr_meta (godotopenxrmeta.gdap + godotopenxrmeta.aar), I suggest we would call this one godot_openxr_khr (godotopenxrkhr.gdap + godotopenxrhkr.aar).

Also right now the meta one builds a -debug.aar and -release.aar but there doesn't seem to be an option in gdap files to export either debug or release versions like we have with GDextensions. Does it make sense to even indifferentiate between the two?

Lastly, should we put all OpenXR loader plugins together in one repo and upload them as a pack to the asset library so it's easy to just install them instead of separating this out?

kisg replied:
@BastiaanOlij If you want to move the building of the KHR plugin to the external repo, then we should just stop building the libopenxr_loader for Android, and add the KHR plugin including the OpenXR Loader sources to the new repo. Maybe we should just use CMake (integrated with Gradle) in that new repo to build the loader so that a single Gradle invocation can build all the plugins. Maybe the in-progress godot_openxr_meta repo should be refactored to hold all the plugins? What do you think?

Regarding debug / release in .gdap: I think this would be a good feature to add, somehow like this:
```
binary=.../currentway.aar
binary_release=.../release.aar
binary_debug=.../debug.aar
```
If binary_release or binary_debug is defined, it takes precedence over binary for the respective build type, but the binary directive could still be available so existing .gdap files could still work.

@m4gr3d What do you think about this?

kisg replied:
> @kisg I'm also in favor of the first option, so let's proceed with that.
> 
> > Should we generate an AAR + gdap file during the generateGodotTemplates task?
> 
> What's the behavior when no loader is provided? Is the logic smart enough to detect that and abort?
> 
> If it's, then the Khronos loader plugin should be generated separately from the Godot templates and provided as an asset as @BastiaanOlij mentioned alongside the other (e.g: Meta) loader plugins.
> 
> If the logic can't handle a missing loader, then it should be generated alongside the Godot template and provided as default.

The logic can handle the missing loader, this was one of the reasons for introducing the dlopen() based approach. So we don't have to provide the loader in the default template.

kisg replied:
Removed the building of openxr_loader on Android

m4gr3d replied:
> Regarding debug / release in .gdap: I think this would be a good feature to add, somehow like this:
> 
> ```
> binary=.../currentway.aar
> binary_release=.../release.aar
> binary_debug=.../debug.aar
> ```
> 
> If binary_release or binary_debug is defined, it takes precedence over binary for the respective build type, but the binary directive could still be available so existing .gdap files could still work.
> 
> @m4gr3d What do you think about this?

@kisg  I think that's a good idea, and it should be fairly straightforward to implement!
Let me know if you have any questions about the way to proceed.

m4gr3d commented on platform/android/java/build.gradle:
```diff
@@ -158,6 +158,8 @@ task zipCustomBuild(type: Zip) {
     }
     from(fileTree(dir: 'app', excludes: ['**/build/**', '**/.gradle/**', '**/*.iml']), fileTree(dir: '.', includes: ['gradlew', 'gradlew.bat', 'gradle/**']))
     include '**/*'
+    from(fileTree(dir: 'lib', includes: ['**/libs/**']))
+    include 'openxr/**/*'
```
@kisg Do we need to include this given that it's not being used by the custom build?

kisg replied:
No, if we agree to provide the Khronos loader as a separate AAR plugin, then we can remove this. Will do it in an update.

BastiaanOlij replied:
I think we pretty much agreed on that so ... :P

kisg replied:
Done.

BastiaanOlij commented:
@kisg @korompg, don't forget to squash the commits into one. Also left a comment on the other PR but I think it should be resolved here but the Vulkan Extension class probably could be changed to use the macros as well

Other than that, @akien-mga I think this can be merged, may need a final ok from the android team?

kisg commented:
@BastiaanOlij Ok, @konczg will do this later today.

BastiaanOlij commented:
@kisg @konczg in the folder `misc/hooks` you can find a number of scripts that you can copy into `.git/hooks` and git will run a bunch of checks on your changes that will correct those style error CI is complaining about.

kisg committed:
Dynamic loading of OpenXR Loader on Android

This change implements dynamic loading of the OpenXR Loader library
on Android. If an OpenXR Loader library is not found,
Godot will still function with OpenXR disabled.

Also, on every platform, the OpenXR symbols are resolved at runtime
using xrGetInstanceProcAddr.

On Windows and Linux the OpenXR loader is included in the main
engine binary.

On Android, the OpenXR Loader is not built with the engine. Separately
distributed Android plugins will be provided with the correct loader
library for each device.

Co-authored-by: Gábor Pál Korom <gabor.p.korom@migeran.com>
Co-authored-by: Gábor Koncz <gabor.koncz@migeran.com>

kisg commented:
@BastiaanOlij @m4gr3d @akien-mga One (hopefully final) update:
- removed the android JNI bindings for the loader, because we are not building the loader for Android anymore
- updated the commit message to reflect the final implementation

Please let me know if we need to change anything to get this merged.

BastiaanOlij reviewed (approved):
Great work, now to get this merged :)

akien-mga commented:
Thanks!
